modules/apps/site/site-cms-site-initializer/src/main/java/com/liferay/site/cms/site/initializer/internal/servlet/DownloadObjectEntryFolderServlet.java

```diff
@@ -0,0 +1,269 @@
+import com.liferay.portal.kernel.servlet.ServletResponseUtil;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.ContentTypes;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.PropsKeys;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.kernel.zip.ZipWriter;
+import com.liferay.portal.kernel.zip.ZipWriterFactory;
+import com.liferay.portal.util.PropsValues;
+
+import jakarta.servlet.Servlet;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Serializable;
+
+import java.util.List;
+import java.util.Map;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Roberto Díaz
+ */
+@Component(
+	property = {
+		"osgi.http.whiteboard.servlet.name=com.liferay.site.cms.site.initializer.internal.servlet.DownloadObjectEntryFolderServlet",
+		"osgi.http.whiteboard.servlet.pattern=/cms/download-folder/*",
+		"servlet.init.httpMethods=GET"
+	},
+	service = Servlet.class
+)
+public class DownloadObjectEntryFolderServlet extends HttpServlet {
+
+	@Override
+	public void service(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException, ServletException {
+
+		_createContext(httpServletRequest, httpServletResponse);
+
+		super.service(httpServletRequest, httpServletResponse);
+	}
+
+	@Override
+	protected void doGet(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException, ServletException {
+
+		try {
+			_downloadObjectEntryFolder(httpServletRequest, httpServletResponse);
+		}
+		catch (PortalException portalException) {
+			throw new ServletException(portalException);
+		}
+	}
+
+	private void _createContext(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		try {
+			EventsProcessorUtil.process(
+				PropsKeys.SERVLET_SERVICE_EVENTS_PRE,
+				PropsValues.SERVLET_SERVICE_EVENTS_PRE, httpServletRequest,
+				httpServletResponse);
+		}
+		catch (ActionException actionException) {
+			if (_log.isDebugEnabled()) {
+				_log.debug(actionException);
+			}
+		}
+	}
+
+	private void _downloadObjectEntryFolder(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException, PortalException {
+
+		String[] pathArray = StringUtil.split(
+			_getObjectEntryFolderPath(httpServletRequest), StringPool.SLASH);
+
+		long classNameId = _portal.getClassNameId(ObjectEntryFolder.class);
+
+		if (classNameId != GetterUtil.getLong(pathArray[0])) {
```

Comment:
@robertoDiaz, can we add a check for the length of pathArray first? We can avoid an exception when calling pathArray[0] if it's and empty array and possible array exceptions later on

Reply:
In this case a Runtime exception is acceptable, I guess. We could not control all the cases.